### PR TITLE
Add fork support to gitlab

### PR DESCRIPTION
**Pull request by AntoineD**

### CODE REVIEW

File: all_repos/push/gitlab_pull_request.py [hunk 1]
```diff
@@ -16,6 +16,7 @@ class Settings(NamedTuple):
     api_key: str
     base_url: str = 'https://gitlab.com/api/v4'
     fork: bool = False
+    https: bool = False
```
Comment (asottile) [inline all_repos/push/gitlab_pull_request.py:19]: this is unrelated -- and I don't want to enable https cloning -- everything should go over ssh

File: all_repos/push/gitlab_pull_request.py [hunk 2]
```diff
@@ -31,14 +31,26 @@ def push(settings: Settings, branch_name: str) -> None:
     }
 
     remote_url = git.remote('.')
-    _, _, repo_slug = remote_url.rpartition(':')
-    repo_slug = _strip_trailing_dot_git(repo_slug)
+    parsed_url = urllib.parse.urlparse(remote_url)
+    repo_slug = _strip_trailing_dot_git(parsed_url.path[1:])
     repo_slug = urllib.parse.quote(repo_slug, safe='')
```
Comment (asottile) [inline all_repos/push/gitlab_pull_request.py:36]: I don't think this logic should have changed

File: tests/push/gitlab_pull_request_test.py [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 import json
 import subprocess
+import urllib
```
Comment (asottile) [inline tests/push/gitlab_pull_request_test.py:5]: importing `urllib` imports nothing -- this needs to actually import the submodules (right now this is depending on import side-effects -- I'm surprised mypy doesn't catch this)

File: tests/push/gitlab_pull_request_test.py [hunk 4]
```diff
@@ -44,9 +45,10 @@ def test_gitlab_pull_request(mock_urlopen, fake_gitlab_repo):
 
     # Pull request should have been made with the commit data
     (req,), _ = mock_urlopen.call_args
+    repo_slug = urllib.parse.quote(fake_gitlab_repo.src.strpath[1:], safe='')
     assert req.get_full_url() == (
         'https://gitlab.com/api/v4/projects'
-        '/user%2Fslug/merge_requests'
+        f'/{repo_slug}/merge_requests'
```
Comment (asottile) [inline tests/push/gitlab_pull_request_test.py:51]: this changes the test -- now it doesn't test anything

File: all_repos/push/gitlab_pull_request.py (hunk 2 above)
Comment (AntoineD) [reply all_repos/push/gitlab_pull_request.py:36]: The original code produced a `repo_slug` that is `//gitlab.com/foo/bar` but we need `foo/bar` (quoting removed for clarity).
See [gitlab API docs](https://docs.gitlab.com/ee/api/merge_requests.html#create-mr), the so called slug can either be an `id` or a [`NAMESPACE/PROJECT_PATH`](https://docs.gitlab.com/ee/user/group/#namespaces).

Also the original code did not work for me.

Comment (asottile) [reply all_repos/push/gitlab_pull_request.py:36]: it shouldn't do that if they are ssh urls?

Comment (AntoineD) [reply all_repos/push/gitlab_pull_request.py:36]: Ah, I see, I'll fix that.

Comment (AntoineD) [reply all_repos/push/gitlab_pull_request.py]: Now only ssh is used.

File: tests/push/gitlab_pull_request_test.py (hunk 3 above)
Comment (AntoineD) [reply tests/push/gitlab_pull_request_test.py]: Fixed.

File: tests/push/gitlab_pull_request_test.py (hunk 4 above)
Comment (AntoineD) [reply tests/push/gitlab_pull_request_test.py]: Fixed.

**Comment by AntoineD** [comment]
In the last commit I removed the usage of the gitlag API in favor of git push options, it works fine on real gitlab projects but the tests fail because the local repo cannot handle push options.
Would you have an idea for a workaround, or maybe we could remove the failing parts?

**Comment by AucaCoyan** [comment]
Hi! Is there an update about this gitlab topic?
Thanks, have a great day!

**Comment by asottile** [comment]
> Hi! Is there an update about the gitlab topic? Thanks, have a great day!

don't do this

there's no comments here so clearly there are no updates

**Comment by AntoineD** [comment]
Hi @asottile, have you seen my question in my [previous comment](https://github.com/asottile/all-repos/pull/239#issuecomment-1184377917)?

**Comment by AntoineD** [comment]
My question was not clear enough, let me please try to rephrase my issue.

The way the tests are done seems no longer adapted because the merge request is no longer created via the gitlab API but instead with git push options.
With the fake repo created for the tests, the call to `git push` via `autofix_lib.run` gives the error `fatal: the receiving end does not support push options\nfatal: the remote end hung up unexpectedly` and I do not know how to adapt the testing in this context.
